Sources/Testing/Running/Configuration.TestFilter.swift

```diff
@@ -178,32 +181,92 @@ extension Configuration.TestFilter {
   }
+}
 
-  /// Initialize this instance to represent a regular expression matched against
-  /// a test's ID.
+// MARK: - Operations
+
+extension Configuration.TestFilter {
+  /// An enumeration which represents filtering logic to be applied to a test
+  /// graph.
+  fileprivate enum Operation {
+    /// A filter operation which has no effect.
+    ///
+    /// All tests are allowed when this operation is applied.
+    case unfiltered
+
+    /// A filter operation which accepts tests included in a precomputed
+    /// selection of test IDs.
+    ///
+    /// - Parameters:
+    ///   - testIDs: The set of test IDs to predicate tests against.
+    ///   - membership: How to interpret the result when predicating tests.
+    case precomputed(_ testIDs: Test.ID.Selection, membership: Membership)
+
+    /// A filter operation which accepts tests which satisfy an arbitrary
+    /// predicate function.
+    ///
+    /// - Parameters:
+    ///   - predicate: The function to predicate tests against.
+    ///   - membership: How to interpret the result when predicating tests.
+    case function(_ predicate: @Sendable (borrowing Test) -> Bool, membership: Membership)
+
+    /// A filter operation which is a combination of other operations.
+    ///
+    /// - Parameters:
+    ///   - lhs: The first test filter's kind.
+    ///   - rhs: The second test filter's kind.
+    ///   - op: The operator to apply when combining the results of the two
+    ///     filters.
+    ///
+    /// The result of applying this filter operation is the combination of
+    /// applying the results of its sub-operations using `op`.
+    indirect case compound(_ lhs: Self, _ rhs: Self, _ op: CombinationOperator)
+  }
+}
+
+extension Configuration.TestFilter.Operation {
+  /// Initialize an instance of this type which implements filtering logic for
+  /// the specified test filter predicate.
   ///
   /// - Parameters:
-  ///   - membership: How to interpret the result when predicating tests.
-  ///   - regex: The regular expression to predicate test IDs against.
+  ///   - predicate: The predicate for which this operation should implement
+  ///     filtering logic.
   ///
-  /// The caller is responsible for ensuring that `regex` is safe to send across
-  /// isolation boundaries. Regular expressions parsed from strings are
-  /// generally sendable.
-  @available(_regexAPI, *)
-  init(membership: Membership, matching regex: Regex<AnyRegexOutput>) {
-    let regex = UncheckedSendable(rawValue: regex)
-    self.init(membership: membership) { test in
-      let id = String(describing: test.id)
-      return id.contains(regex.rawValue)
+  /// - Throws: Any error encountered while converting the specified predicate
+  ///   into an operation. One example is the creation of a `Regex` from a
+  ///   `.pattern` predicate: if the pattern is not a valid regular expression,
+  ///   an error will be thrown.
+  init(_ predicate: Configuration.TestFilter.Predicate) throws {
```

Comment:
`func makeOperation()` on `TestFilter` maybe? 🤔 The initializer syntax would suggest it's a conversion/cast, but it's more of a factory function IMHO.